pkg/virt-launcher/virtwrap/network/ndp/connection.go

```diff
@@ -0,0 +1,126 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package ndp
+
+import (
+	"fmt"
+	"net"
+	"os"
+
+	"github.com/mdlayher/ndp"
+	"golang.org/x/net/ipv6"
+)
+
+const (
+	chkOff  = 2
+	maxHops = 255
+)
+
+// A NDPConnection instruments a system.Conn and adds retry functionality for
+// receiving / sending NDP messages on a given interface.
+type NDPConnection struct {
+	iface      *net.Interface
+	rawConn    *net.IPConn
+	conn       *ipv6.PacketConn
+	controlMsg *ipv6.ControlMessage
+}
+
+// Return an NDPConnection bound to the chosen interface.
+func NewNDPConnection(ifaceName string) (*NDPConnection, error) {
+	iface, err := net.InterfaceByName(ifaceName)
+	if err != nil {
+		return nil, fmt.Errorf("could not find interface %s: %v", ifaceName, err)
+	}
+
+	listenAddr := &net.IPAddr{
+		IP:   net.IPv6unspecified,
+		Zone: ifaceName,
+	}
+	icmpListener, err := net.ListenIP("ip6:ipv6-icmp", listenAddr)
+	if err != nil {
+		return nil, fmt.Errorf("could not listen to ip6:ipv6-icmp on addr %s: %v", listenAddr.String(), err)
+	}
+
+	ipv6Conn := ipv6.NewPacketConn(icmpListener)
+
+	_ = ipv6Conn.SetHopLimit(maxHops)          // as per RFC 4861, section 4.1
+	_ = ipv6Conn.SetMulticastHopLimit(maxHops) // as per RFC 4861, section 4.1
+
+	// Calculate and place ICMPv6 checksum at correct offset in all messages.
+	if err := ipv6Conn.SetChecksum(true, chkOff); err != nil {
+		return nil, fmt.Errorf("could not enable ICMPv6 checksum processing: %v", err)
+	}
+
+	routersMulticastGroup := &net.IPAddr{
+		IP:   net.IPv6linklocalallrouters,
+		Zone: ifaceName,
+	}
+	if err := ipv6Conn.JoinGroup(iface, routersMulticastGroup); err != nil {
+		return nil, fmt.Errorf("failed to join %s multicast group: %v", routersMulticastGroup.String(), err)
+	}
+
+	listener := &NDPConnection{
+		iface:   iface,
+		conn:    ipv6Conn,
+		rawConn: icmpListener,
+		controlMsg: &ipv6.ControlMessage{
+			HopLimit: maxHops,
```

Comment:
AFAIU, this control message is used for to outgoing packets.
Why the `HopLimit` and the `Src` should be part of the control message? Could find in the RFC a reference the RA message should contain those feilds in the control message.

Regarding the `IfIndex`, I guess it is needed for the same reasons it is need in dhcpv6 (and v4). So that the outgoing interface will be the bridge one.

But as I wrote [here](https://github.com/kubevirt/kubevirt/pull/4581/commits/5b9b33995d8bd0f0449830c4992739ea7c90bd85#r553302183) about the incoming packets, maybe since we join to the multicast group on top of the bridge, it is not needed?